zcash_client_sqlite/src/wallet/init/migrations/tx_retrieval_queue.rs

```diff
@@ -0,0 +1,83 @@
+//! A migration to add the `tx_retrieval_queue` table to the database.
+
+use rusqlite::{named_params, Transaction};
+use schemer_rusqlite::RusqliteMigration;
+use std::collections::HashSet;
+use uuid::Uuid;
+use zcash_primitives::transaction::builder::DEFAULT_TX_EXPIRY_DELTA;
+
+use crate::wallet::init::WalletMigrationError;
+
+use super::utxos_to_txos;
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0xfec02b61_3988_4b4f_9699_98977fac9e7f);
+
+pub(crate) struct Migration;
+
+impl schemer::Migration for Migration {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [utxos_to_txos::MIGRATION_ID].into_iter().collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Adds a table for tracking transactions to be downloaded for transparent output and/or memo retrieval."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            "CREATE TABLE tx_retrieval_queue (
+                txid BLOB NOT NULL UNIQUE,
+                query_type INTEGER NOT NULL,
+                dependent_transaction_id INTEGER,
+                FOREIGN KEY (dependent_transaction_id) REFERENCES transactions(id_tx)
+            );
+
+            ALTER TABLE transactions ADD COLUMN target_height INTEGER;
+
+            CREATE TABLE transparent_spend_search_queue (
+                address TEXT NOT NULL,
+                transaction_id INTEGER NOT NULL,
+                output_index INTEGER NOT NULL,
+                FOREIGN KEY (transaction_id) REFERENCES transactions(id_tx),
+                CONSTRAINT value_received_height UNIQUE (transaction_id, output_index)
+            );",
+        )?;
+
+        transaction.execute(
+            "UPDATE transactions
+             SET target_height = expiry_height - :default_expiry_delta
+             WHERE expiry_height IS NOT NULL AND created IS NOT NULL",
+            named_params![":default_expiry_delta": DEFAULT_TX_EXPIRY_DELTA],
+        )?;
+
+        Ok(())
+    }
+
+    fn down(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            "DROP TABLE transparent_spend_search_queue;
+             ALTER TABLE transactions DROP COLUMN target_height;
+             DROP TABLE tx_retrieval_queue;",
+        )?;
```

Comment:
When do we care about being able to revert a migration, and when do we not?

Reply:
We don't currently have any code paths that exercise down migrations. We should perhaps universally not implement them; we would always implement a fix to a previous migration as a forward migration, rather than trying to down-migrate.